sled-agent/src/storage_manager.rs

```diff
@@ -105,131 +116,379 @@ impl Pool {
+        Self {
+            pool_name: pool_name.to_string(),
+            dataset_name: dataset_name.to_string(),
+        }
+    }
+
+    fn full(&self) -> String {
+        format!("{}/{}", self.pool_name, self.dataset_name)
+    }
 }
 
 // Description of a dataset within a ZFS pool, which should be created
 // by the Sled Agent.
-struct PartitionInfo<'a> {
-    name: &'a str,
-    zone_prefix: &'a str,
-    data_directory: &'a str,
-    svc_directory: &'a str,
-    port: u16,
+#[derive(Clone, Debug, Deserialize, Serialize, JsonSchema)]
+struct DatasetInfo {
+    // TODO: Is this always "/data"?
+    data_directory: String,
+    address: SocketAddr,
     kind: DatasetKind,
+    name: DatasetName,
 }
 
+impl DatasetInfo {
+    fn new(pool: &str, kind: DatasetKind, address: SocketAddr) -> DatasetInfo {
+        match kind {
+            DatasetKind::CockroachDb { .. } => DatasetInfo {
+                name: DatasetName::new(pool, "cockroachdb"),
+                data_directory: "/data".to_string(),
+                address,
+                kind,
+            },
+            DatasetKind::Clickhouse { .. } => DatasetInfo {
+                name: DatasetName::new(pool, "clickhouse"),
+                data_directory: "/data".to_string(),
+                address,
+                kind,
+            },
+            DatasetKind::Crucible { .. } => DatasetInfo {
+                name: DatasetName::new(pool, "crucible"),
+                data_directory: "/data".to_string(),
+                address,
+                kind,
+            },
+        }
+    }
+
+    fn zone_prefix(&self) -> String {
+        format!("{}{}_", ZONE_PREFIX, self.name.full())
+    }
+
+    async fn start_zone(
+        &self,
+        log: &Logger,
+        zone: &RunningZone,
+        address: SocketAddr,
+        do_format: bool,
+    ) -> Result<(), Error> {
+        match self.kind {
+            DatasetKind::CockroachDb { .. } => {
+                // Load the CRDB manifest.
+                zone.run_cmd(&[
+                    crate::illumos::zone::SVCCFG,
+                    "import",
+                    "/var/svc/manifest/site/cockroachdb/manifest.xml",
+                ])?;
+
+                // Set parameters which are passed to the CRDB binary.
+                zone.run_cmd(&[
+                    crate::illumos::zone::SVCCFG,
+                    "-s",
+                    "svc:system/illumos/cockroachdb",
+                    "setprop",
+                    &format!("config/listen_addr={}", address),
+                ])?;
+                zone.run_cmd(&[
+                    crate::illumos::zone::SVCCFG,
+                    "-s",
+                    "svc:system/illumos/cockroachdb",
+                    "setprop",
+                    &format!("config/store={}", self.data_directory),
+                ])?;
+                // TODO: Set these addresses, use "start" instead of
+                // "start-single-node".
```

Comment:
:clock1030:

Reply:
Good reminder, I'll add a note to https://github.com/oxidecomputer/omicron/issues/727 about NTP